packages/go_router/test/go_router_test.dart

```diff
@@ -882,6 +896,51 @@ void main() {
       expect(router.location, '/login');
     });
 
+    testWidgets('top-level redirect take priority over route level',
+        (WidgetTester tester) async {
+      final List<GoRoute> routes = <GoRoute>[
+        GoRoute(
+          path: '/',
+          builder: (BuildContext context, GoRouterState state) =>
+              const HomeScreen(),
+          routes: <GoRoute>[
+            GoRoute(
+                path: 'dummy',
+                builder: (BuildContext context, GoRouterState state) =>
+                    const DummyScreen(),
+                redirect: (GoRouterState state) => '/dummy2'),
+            GoRoute(
+                path: 'dummy2',
+                builder: (BuildContext context, GoRouterState state) =>
+                    const DummyScreen()),
+            GoRoute(
+                path: 'login',
+                builder: (BuildContext context, GoRouterState state) =>
+                    const LoginScreen()),
+          ],
+        ),
+      ];
+
+      final GoRouter router = await createRouter(routes, tester,
+          redirect: (GoRouterState state) =>
+              state.subloc == '/dummy' || state.subloc == '/login'
+                  ? null
+                  : '/login');
```

Comment:
Could we make this throw an exception instead since this code is expected to never run?